corehq/util/couch_doc_processor.py

```diff
@@ -29,162 +52,34 @@ class ResumableDocsByTypeIterator(object):
-                retried[doc_id] = retries
-                try:
-                    yield self.db.get(doc_id)
-                except ResourceNotFound:
-                    pass
-
-        # save iteration state without offset to signal completion
-        self.state.pop("offset", None)
-        self.state["retry"] = {}
-        self._save_state()
-
-    def retry(self, doc, max_retry=3):
-        """Add document to be yielded at end of iteration
-
-        Iteration order of retry documents is undefined. All retry
-        documents will be yielded after the initial non-retry phase of
-        iteration has completed, and every retry document will be
-        yielded each time the iterator is stopped and resumed during the
-        retry phase. This method is relatively inefficient because it
-        forces the iteration state to be saved to couch. If you find
-        yourself calling this for many documents during the iteration
-        you may want to consider a different retry strategy.
-
-        :param doc: The doc dict to retry. It will be re-fetched from
-        the database before being yielded from the iteration.
-        :param max_retry: Maximum number of times a given document may
-        be retried.
-        :raises: `TooManyRetries` if this method has been called too
-        many times with a given document.
-        """
-        doc_id = doc["_id"]
-        retries = self.state["retry"].get(doc_id, 0) + 1
-        if retries > max_retry:
-            raise TooManyRetries(doc_id)
-        self.state["retry"][doc_id] = retries
-        self._save_state()
+    doc_types = sorted(doc_types)
+    data_function.__name__ = " ".join(doc_types)
 
-    @property
-    def progress_info(self):
-        """Extra progress information
+    view_kwargs = {
+        'limit': chunk_size,
+        'include_docs': True,
+        'reduce': False,
+        'startkey': [doc_types[0]],
+        'endkey': [doc_types[0], {}]
+    }
 
-        This property can be used to store and retrieve extra progress
-        information associated with the iteration. The information is
-        persisted with the iteration state in couch.
-        """
-        return self.state.get("progress_info")
+    args_provider = ResumableDocsByTypeArgsProvider(view_kwargs, doc_types)
 
-    @progress_info.setter
-    def progress_info(self, info):
-        self.state["progress_info"] = info
-        self._save_state()
-
-    def _save_state(self):
-        self.state["timestamp"] = datetime.utcnow().isoformat()
-        self.db.save_doc(self.state)
+    class ResumableDocsIterator(ResumableFunctionIterator):
+        def __iter__(self):
+            for result in super(ResumableDocsIterator, self).__iter__():
+                yield result['doc']
 
-    def discard_state(self):
+    def item_getter(doc_id):
         try:
-            self.db.delete_doc(self.iteration_id)
+            return {'doc': db.get(doc_id)}
```

Comment:
Not understanding this, and I can't find the place where this gets unwrapped. I thought the iterator was always dealing with doc dicts, not `{"doc": the_real_doc}`. I assume you're using the wrapper as a signal in case `db.get(doc_id)` returns `None`? Seems like it would be better to this:
```py
try:
    return db.get(doc_id)
except ResourceNotFound:
    raise SkipRetry
```